pennylane_lightning/core/src/simulators/lightning_gpu/utils/DataBuffer.hpp

```diff
@@ -0,0 +1,229 @@
+            PL_CUDA_IS_SUCCESS(cudaGetDevice(&local_dev_id));
+
+            length_ = other.length_;
+            dev_tag_ =
+                DevTag<DevTagT>{local_dev_id, other.dev_tag_.getStreamID()};
+            dev_tag_.refresh();
+            PL_CUDA_IS_SUCCESS(
+                cudaMalloc(reinterpret_cast<void **>(&gpu_buffer_),
+                           sizeof(GPUDataT) * length_));
+            CopyGpuDataToGpu(other.gpu_buffer_, other.length_);
+        }
+        return *this;
+    }
+
+    DataBuffer &operator=(DataBuffer &&other) {
+        if (this != &other) {
+            int local_dev_id = -1;
+            PL_CUDA_IS_SUCCESS(cudaGetDevice(&local_dev_id));
+            length_ = other.length_;
+            if (local_dev_id == other.dev_tag_.getDeviceID()) {
+                dev_tag_ = std::move(other.dev_tag_);
+                dev_tag_.refresh();
+
+                gpu_buffer_ = other.gpu_buffer_;
+            } else {
+                dev_tag_ =
+                    DevTag<DevTagT>{local_dev_id, other.dev_tag_.getStreamID()};
+                dev_tag_.refresh();
+
+                PL_CUDA_IS_SUCCESS(
+                    cudaMalloc(reinterpret_cast<void **>(&gpu_buffer_),
+                               sizeof(GPUDataT) * length_));
+                CopyGpuDataToGpu(other.gpu_buffer_, other.length_);
+                PL_CUDA_IS_SUCCESS(cudaFree(other.gpu_buffer_));
+                other.dev_tag_ = {};
+            }
+            other.length_ = 0;
+            other.gpu_buffer_ = nullptr;
+        }
+        return *this;
+    };
+
+    virtual ~DataBuffer() {
+        if (gpu_buffer_ != nullptr) {
+            PL_CUDA_IS_SUCCESS(cudaFree(gpu_buffer_));
+        }
+    };
+
+    /**
+     * @brief Zero-initialize the GPU buffer.
+     *
+     */
+    void zeroInit() {
+        PL_CUDA_IS_SUCCESS(
+            cudaMemset(gpu_buffer_, 0, length_ * sizeof(GPUDataT)));
+    }
+
+    auto getData() -> GPUDataT * { return gpu_buffer_; }
+    auto getData() const -> const GPUDataT * { return gpu_buffer_; }
+    auto getLength() const { return length_; }
+
+    /**
+     * @brief Get the CUDA stream for the given object.
+     *
+     * @return const cudaStream_t&
+     */
+    inline auto getStream() const -> cudaStream_t {
+        return dev_tag_.getStreamID();
+    }
+
+    inline auto getDevice() const -> int { return dev_tag_.getDeviceID(); }
+
+    inline auto getDevTag() const -> const DevTag<DevTagT> & {
+        return dev_tag_;
+    }
+
+    /**
+     * @brief Copy data from another GPU memory block to here.
+     *
+     */
+    void CopyGpuDataToGpu(const GPUDataT *gpu_in, std::size_t length,
+                          bool async = false) {
+        PL_ABORT_IF_NOT(
+            getLength() == length,
+            "Sizes do not match for GPU data. Please ensure the source "
+            "buffer is not larger than the destination buffer");
+        if (async) {
+            PL_CUDA_IS_SUCCESS(cudaMemcpyAsync(
```

Comment:
We tend not to use these often, as it requires additional changes in the pipeline (better stream supports across the package). I think we can just accept these aren't tested for now, and maybe hide them from coverage reports --- unless an explicit set of tests are added, which would alleviate the issue at a unit level, rather than within integration.